Add sierra assertion when tracking sierra gas

THenry14 opened the pull request:
Related to #2977

This PR stack aims to add support for new resource tracking (sierra-gas) and new gas representation (GasVector triplet instead of single l1 gas value).

-- (https://github.com/foundry-rs/starknet-foundry/pull/3049) Add sierra gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3050) Run already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add support for sierra gas in `--save-trace-data`
--> (This PR) Add sierra version assertion
-- (https://github.com/foundry-rs/starknet-foundry/pull/3054) Introduce new gas representation
-- (https://github.com/foundry-rs/starknet-foundry/pull/3055) Add gas tests for sierra-gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3056) Update docs with sierra-gas related changes

THenry14 committed:
[sip-ci] Add sierra assertion when tracking sierra gas

ddoktorski reviewed (commented):
Shouldn't it be sufficient to verify that contracts were compiled with `Scarb >=2.10.0` rather than checking Sierra version explicitly?

THenry14 commented:
> Shouldn't it be sufficient to verify that contracts were compiled with `Scarb >=2.10.0` rather than checking Sierra version explicitly?

Yes, it could be done this way. 
I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.

If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)

cptartur commented on crates/forge/src/compatibility_check.rs:
```diff
@@ -115,10 +117,80 @@ fn get_raw_version(name: &str, command: &RefCell<Command>) -> Result<String> {
+            let contract_version = parse_sierra_version(sierra_str)?;
+            if MINIMAL_SIERRA_VERSION_FOR_SIERRA_GAS > contract_version {
+                bail!(
+                    "Contract version {} is lower than required minimal sierra version",
```
Can we do something along the lines of this?

```suggestion
                    "Contract version {} is lower than required minimal sierra version {MINIMAL_SIERRA_VERSION_FOR_SIERRA_GAS}",
```

cptartur commented on crates/forge/src/compatibility_check.rs:
```diff
@@ -115,10 +117,80 @@ fn get_raw_version(name: &str, command: &RefCell<Command>) -> Result<String> {
+
+    let parsed_values: Vec<u8> = sierra_json["sierra_program"]
+        .as_array()
+        .context("Unable to read `sierra_program`. Ensure it is an array of felts.")?
```
This is not user generated so it cannot be fixed by a user anyway.

```suggestion
        .context("Unable to read `sierra_program`.")?
```

cptartur commented on crates/forge/src/compatibility_check.rs:
```diff
@@ -115,10 +117,80 @@ fn get_raw_version(name: &str, command: &RefCell<Command>) -> Result<String> {
+    if forge_config.test_runner_config.tracked_resource == ForgeTrackedResource::SierraGas {
+        verify_contracts_sierra_versions(forge_config).with_context(||
+            format!(
+                "Tracking SierraGas is not supported for sierra <= {MINIMAL_SIERRA_VERSION_FOR_SIERRA_GAS}"
```
This is kind of duplicating the error you get from the inside of the function.

THenry14 replied:
Not really, it adds context - what I wanted to achieve? - track sierra. what was the error? (inner function - wrong sierra version).

cptartur replied:
Hmm, okay, but for me it looks kind of verbose

cptartur commented on crates/forge/src/compatibility_check.rs:
```diff
@@ -249,4 +321,116 @@ mod tests {
+                execution_data_to_save: ExecutionDataToSave::default(),
+            }),
+        };
+        assert!(check_sierra_gas_version_requirement(&forge_config).is_err());
```
Can we do `unwrap_err` and assert the error contents?

THenry14 replied:
Added, please note that there is also a e2e test that asserts this `sierra_gas_with_older_scarb`

ksew1 commented:
> Yes, it could be done this way.
I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.

> If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)

To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.

THenry14 committed:
review changes

THenry14 commented:
> > Yes, it could be done this way.
> > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> 
> > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> 
> To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.

I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts.
Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔

cptartur commented:
> > > Yes, it could be done this way.
> > > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> > 
> > 
> > > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> > 
> > 
> > To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.
> 
> I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts. Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔

Is it possible for new Scarb versions to still generate older sierra somehow? If yes, we should check that. But otherwise it may be a bit overkill.

cptartur commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/contracts_data.rs:
```diff
@@ -105,6 +105,11 @@ impl ContractsData {
         self.selectors.get(entry_point_selector)
     }
+
+    #[must_use]
+    pub fn get_contract_names(&self) -> Option<Vec<&ContractName>> {
+        (!self.contracts.is_empty()).then(|| self.contracts.keys().collect())
+    }
```
I didn't understand what this code does correctly

THenry14 commented:
> > Yes, it could be done this way.
> > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> 
> > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> 
> To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.

I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts.
Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔 

> > > > Yes, it could be done this way.
> > > > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> > > 
> > > 
> > > > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> > > 
> > > 
> > > To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.
> > 
> > 
> > I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts. Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔
> 
> Is it possible for new Scarb versions to still generate older sierra somehow? If yes, we should check that. But otherwise it may be a bit overkill.

Scarb is always tied up to a specific cairo + sierra versions, so there is no way an older scarb suddenly starts supporting newer sierra and the other way around.

ksew1 commented:
> > > > > Yes, it could be done this way.
> > > > > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> > > > 
> > > > 
> > > > > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> > > > 
> > > > 
> > > > To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.
> > > 
> > > 
> > > I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts. Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔
> > 
> > 
> > Is it possible for new Scarb versions to still generate older sierra somehow? If yes, we should check that. But otherwise it may be a bit overkill.
> 
> Scarb is always tied up to a specific cairo + sierra versions, so there is no way an older scarb suddenly starts supporting newer sierra and the other way around.

If I had to leak my personal opinion, I would likely prefer the "fail-fast" model.  

By making assertions at the start of the program, we can detect issues early rather than encountering errors halfway through execution. This will improve maintainability since all checks are clearly defined in one place, making it easier to understand how they relate to each other.  

Also, I would base the check purely on Scarb version. Most users are probably not familiar with Sierra or how it relates to Scarb, so keeping it Scarb-centric would make it more accessible.

cptartur commented:
> > > > > > Yes, it could be done this way.
> > > > > > I have opted, however, with checking contract class data directly because I wanted to be explicit - after all I don't care about scarb per se, but about sierra; I already have the contract class loaded, it's just a matter of parsing it for version, which is not that expensive.
> > > > > 
> > > > > 
> > > > > > If you have strong opinion on this let's discuss, personally I don't see many advantages from checking scarb version over this, but am ready to be schooled :)
> > > > > 
> > > > > 
> > > > > To be honest, I think we should check for `scarb` versions. It is way easier for the user to understand that they have the wrong `scarb` version and should change it, rather than realizing they have the wrong `sierra` version.
> > > > 
> > > > 
> > > > I see your point 🤔 But I also really think this should be done on package level, when we actually need the artifacts. Do you think it would be better, if the error also stated minimal scarb version? (so not checking scarb per se, but tying it when interfacing with the user). That way user would be able to quickly tell which scarb version is needed, but we still would be able to check this per artifact. Maybe it's a stretch though. 🤔
> > > 
> > > 
> > > Is it possible for new Scarb versions to still generate older sierra somehow? If yes, we should check that. But otherwise it may be a bit overkill.
> > 
> > 
> > Scarb is always tied up to a specific cairo + sierra versions, so there is no way an older scarb suddenly starts supporting newer sierra and the other way around.
> 
> If I had to leak my personal opinion, I would likely prefer the "fail-fast" model.
> 
> By making assertions at the start of the program, we can detect issues early rather than encountering errors halfway through execution. This will improve maintainability since all checks are clearly defined in one place, making it easier to understand how they relate to each other.
> 
> Also, I would base the check purely on Scarb version. Most users are probably not familiar with Sierra or how it relates to Scarb, so keeping it Scarb-centric would make it more accessible.

Sierra information could be relevant to the users that are aware of the USC existing, so maybe it can still be printed. But definitely an information about Scarb version is required.

THenry14 committed:
Replace sierra check with scarb check

THenry14 commented:
Yeah, You're right guys, I'm convinced :+1:
Made the necessary changes:
- we now check scarb version, and not sierra version
- the check is moved to a requirements checker (as it makes more sense now to verify pre-compilation)
- Info about sierra is still retained in error messsage, as scarb is tied to specific sierra version

THenry14 committed:
New gas representation (#3054)

Related to #2977

This PR stack aims to add support for new resource tracking (sierra-gas)
and new gas representation (GasVector triplet instead of single l1 gas
value).

-- (https://github.com/foundry-rs/starknet-foundry/pull/3049) Add sierra
gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3050) Run
already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add
support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add
support for sierra gas in `--save-trace-data`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3053) Add sierra
version assertion
--> (This PR) Introduce new gas representation
-- (https://github.com/foundry-rs/starknet-foundry/pull/3055) Add gas
tests for sierra-gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3056) Update
docs with sierra-gas related changes
